### PR TITLE
Terminate indexing pipeline once all merge operations have been done

### PR DESCRIPTION
**Pull request by fmassot**
Accurate termination condition:
- source + indexer + packager + uploader + publisher have exited successfully
- all actors of the merge pipeline are in `IDLE` state

## Test
I need to test it on a real dataset. For the moment, I did not find a nice way to add a unit test on the pipeline termination :/

### CODE REVIEW

File: quickwit-indexing/src/actors/pipeline_supervisor.rs [hunk 1]
```diff
@@ -416,7 +414,40 @@ impl IndexingPipelineSupervisor {
...
+                        // This is true when all merging pipeline actors are all in `IDLE` state
+                        // and indexing actors have exited.
+                        if handlers.source.state() == ActorState::Exit
+                            && handlers.indexer.state() == ActorState::Exit
+                            && handlers.packager.state() == ActorState::Exit
+                            && handlers.uploader.state() == ActorState::Exit
+                            && handlers.publisher.state() == ActorState::Exit
+                            && handlers.merge_planner.state() == ActorState::Idle
+                            && handlers.merge_split_downloader.state() == ActorState::Idle
+                            && handlers.merge_executor.state() == ActorState::Idle
```
Comment (fulmicoton) [inline quickwit-indexing/src/actors/pipeline_supervisor.rs:430]: I think there is a small race condition. (Maybe we don't care).
There are many ways for this condition to be true at one point of time, and not true later.

Comment (fulmicoton) [reply quickwit-indexing/src/actors/pipeline_supervisor.rs:430]: Again better than nothing, but I think it is important to be aware of that.

Comment (fmassot) [reply quickwit-indexing/src/actors/pipeline_supervisor.rs:430]: Indeed, I was not very confident when I wrote this code, many equality checks on shifting sands.
I will leave it a few days and see if I come up with better ideas or at least some docs to explain the limitations of this approach.

Comment (fmassot) [reply quickwit-indexing/src/actors/pipeline_supervisor.rs:430]: To make the shutdown more reliable, here is what we can do when there is only the merge pipeline running:
- keep a boolean to track the fact that the merge pipeline was idle
- on supervise
  - if merge pipeline is idle then: if bool is true, exit if not set bool to true
  - if not, set bool to false

Comment (fmassot) [reply quickwit-indexing/src/actors/pipeline_supervisor.rs]: @fulmicoton I added a state variable on the indexing pipeline to terminate the pipeline after 2 consecutive idle merge pipeline states.

**Comment by fmassot** [comment]
Closing this PR for now.

**Pull request closed without merging**